### DISCUSSION TITLE
Testing AES hardware acceleration support using the latest openssl versions 3.1.1 or higher

### DISCUSSION THREAD

**Discussion by appapurapu** [openssl/openssl#21710] [Q&A]
We have c++ based cross platform applications(Linux, Windows, Android, web, etc) linked with openssl static libraries.
I am trying to test the hardware acceleration support using openssl EVP_* apis, but I am unable to find one which finds it.
In earlier releases, we can see couple of APIs like OPENSSL_ia32cap_loc() or CRYPTO_get_hw_acceleration_enabled() or EVP_has_aes_hardware(), etc. exists but not in the latest releases.
Are there any openssl public APIs exists to determine the AES hardware acceleration support? If not, what is the best approach for the c++ applications (with static openssl libraries) to test aes hardware acceleration support ?
is the latest OpenSSL 3.1.1 or higher versions uses the CPUID instruction to determine AES hardware acceleration support?

Thanks for your time reading this question.
Best Regards,
RajaA

**Comment by nhorman** [comment]
I _think_ (emphasis on the think, I'm just getting started looking through the source), that there is no explicit inherent support for hardware acceleration in the core openssl api (at least not in the 3+ branches).  There appears to be some opportunistic use of assembly insturction that implement some accelerated operations when the base arch that you are building for supports them, but if you're look for support for things like the intel [cryptographic intrinsics](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#cats=Cryptography) or [ARM neon intrinsics](https://developer.arm.com/architectures/instruction-sets/intrinsics/#f:@navigationhierarchiessimdisa=[Neon]&f:@navigationhierarchiesinstructiongroup=[Cryptography]), you're not likely to find them.

As I understand it, the proscribed method of implementation to use code like that is through the [provider api](https://www.openssl.org/docs/manmaster/man7/provider.html).  That is to say, you write a provider dso that implements the algorithms you are interested in and when registered and selected, the provider hook use the accelerated methods you are interested in leveraging.  Those don't currently ship with openssl, you can find a list of the included providers[ here](https://github.com/openssl/openssl/blob/master/README-PROVIDERS.md).  You can also find providers that others have written on git hub (like this pkcs11 [provider](https://github.com/latchset/pkcs11-provider).  Thats not exactly what you want of course, but it will give you an idea of how to do what you are looking for.

As for how to detect the availability of those accelerated instructions, that gets rolled into the query function of the provider.  I would use something like gcc's __builtin_cup_supports intrinsic to do a runtime check for cpu support for any instructions you want to leverage, and return the appropriate code based on that runtime determination

Thats addresses intrinsic cpu support, Theres also the possibilty that you might be using off cpu hardware cryptographic acceleration (for instance intel QuickAssist hardware).  Support for these devices is offered in a myriad of ways, for which you would also may need to write a provider.  For instance, with QAT, intel offers a library that you can link to that addresses the hardware via VFIO.  for that you would write a provider that interfaces to that library.  Intel also includes support for QAT through the kernel via the AF_ALG protocol.  In that case, you might be in luck, as I think openssl offers an AF_ALG provider that could interface to it.  I've not used that provider (or engine when I used it with openssl 1.1), but at the time it worked fairly well.

Hope that helps.

**Comment by paulidale** [comment]
There is no way to directly use OpenSSL's CPU feature detection code.  All of the EVP algorithm interfaces will use HW acceleration if available (& supported by the version of OpenSSL you're using).  The deprecated low level API might or might not depending on which is called.

The best approach for your application to detect CPU features is to get the CPUID information and interpret the results yourself. I.e. do not rely on OpenSSL to give you this information.

If you set [`OPENSSL_ia32cap`](https://www.openssl.org/docs/manmaster/man3/OPENSSL_ia32cap.html) you can override the internal CPU feature detection.  The environment variable is only checked once when first needed and cannot give you features that your CPU is lacking.

**Comment by appapurapu** [comment]
Thanks for all the suggestions. 
This information is good for Linux / Windows assuming they are using Intel chipset. Using CPUID info is the way to go.
What about armv8 architecture based android devices?

**Reply by paulidale** [reply]
The [openssl-env documentation](https://www.openssl.org/docs/manmaster/man7/openssl-env.html) lists all that are available.  For aarch64 use `OPENSSL_armcap` but the specific meaning of bits isn't currently documented outside the source code.

**Comment by nhorman** [comment]
most compilers also have intrinsics which just let you test cpu capabilities support natively.  For arm on linux there is the getauxval(AT_HWCAP | AT_HWCAP2) call which returns the arm auxv vector which you can test for various capabilities, and the os will include support definitions for (see HWCAP_NEON, as an example)
